### PR TITLE
Add LIST command to agent

### PR DESCRIPTION
**Pull request by mcncl**
## Changes
This implements a `list` command on the `agent` that lists all agents within a users organisation.

I've also added support for setting the `--output|-o` to either JSON or YAML. See below.

bk_list_1

bk_list_2

### Update
I've removed the table printing for now; the sensible default feels like JSON as folks can actually do somethting with the returned data.

We can add the table back in if we want to, but other than a visual thing it serves no purpose and would actually be more difficult to use; with JSON we could use `jq` to use the returned data, for example.

CleanShot 2023-09-18 at 09 51 44

### Additional Update
* I've refactored the `print` functions to accept anything so they can be re-used
* I've created a single `view` command which takes in an `--id` and then produces output of the given agent
* Commands, where applicable, now all accept an `--output|-o` flag to set either `json` or `yaml`. Default is currently `json`

CleanShot 2023-09-18 at 17 17 09

### CODE REVIEW

File: pkg/cmd/agent/list.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+		DisableFlagsInUseLine: true,
+		Use:                   "list",
+		Args:                  cobra.NoArgs,
+		Short:                 "Lists the agents for the current organization",
+		Long: heredoc.Doc(`
+            Command to list all agents for the current organization.
+
+
+            Only running agents are listed by default. To list all agents, use the
+            --all flag.
```
Comment (jradtilbrook) [inline pkg/cmd/agent/list.go:22]: Remove this since there is no flag?

File: pkg/cmd/agent/list.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+		`),
+		RunE: func(cmd *cobra.Command, args []string) error {
+            agents, _, err := f.RestAPIClient.Agents.List(f.Config.Organization, &buildkite.AgentListOptions{})
+
+            for i, agent := range agents {
+                fmt.Printf("%d: %s. ID: %s. State: %s\n", i+1, *agent.Name, *agent.ID, *agent.ConnectedState)
```
Comment (jradtilbrook) [inline pkg/cmd/agent/list.go:28]: I think there is a table output format but this will do for now 👍

**Review by jradtilbrook — CHANGES_REQUESTED** [review]

File: internal/printer/printer.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+)
+
+type Output string
+
+const (
+	JSON Output = "json"
+	YAML Output = "yaml"
+)
+
+func PrintOutput(output Output, p any) (string, error) {
```
Comment (jradtilbrook) [inline internal/printer/printer.go:16]: In terms of code maintenance, I'd like to see this abstracted more and configured as part of startup. So something early on from the root command could read the output flag and set a global config somewhere. That way its not repeated in individual commands and those commands just call a basic "print" type function that knows what format. 

But another thing to refactor for now

File: pkg/cmd/agent/agent.go [hunk 4]
```diff
@@ -6,11 +6,18 @@ import (
...
 	"github.com/buildkite/cli/v3/pkg/cmd/validation"
+	"github.com/buildkite/go-buildkite/v3/buildkite"
 	"github.com/spf13/cobra"
 )
 
+type Agent buildkite.Agent
+
+func (a Agent) Print(output printer.Output) {
+}
+
```
Comment (jradtilbrook) [inline pkg/cmd/agent/agent.go:20]: Was this the beginning of something you've abandoned now?

File: pkg/cmd/agent/view.go [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+			data, err := printer.PrintOutput(printer.Output(output), agent)
+			if err != nil {
+				return err
+			}
+			fmt.Println(data)
+			return nil
+		},
+	}
+
+	cmd.Flags().String("id", "", "The ID of the agent")
```
Comment (jradtilbrook) [inline pkg/cmd/agent/view.go:45]: I don't think this should be a flag, it should be an arg? eg [`gh issue view <num>`](https://primer.style/cli/foundations/language)

File: pkg/cmd/validation/uuid.go [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package validation
+
+import "github.com/google/uuid"
+
+// UUID validates that the given string is a valid UUID.
+func ValidateUUID(s string) error {
```
Comment (jradtilbrook) [inline pkg/cmd/validation/uuid.go:6]: I'm not sure the value in this, either we'll get an error from the API or can validate it early.

But I think it is useful for users to be able to use the agent URL as well and we parse that out. So they can easily copy/paste from the web

**Review by jradtilbrook — CHANGES_REQUESTED** [review]
Looking really good

I like how we have json and yaml options now 👌  But I disagree on json being the default. I think defaulting to a table/more human readable/consumable format is better and then if users are going to make use of this in some automation, they will add the -o json flag. But we can follow up with that later.

I've left a few comments here but most we can follow up with in another PR. The only one I'd like to see changed now is the `--id` flag become an argument instead

File: pkg/cmd/agent/view.go (hunk 5 above)
Comment (mcncl) [reply pkg/cmd/agent/view.go:45]: Depends entirely on which CLI you're trying to replicate; in `aws` it's a flag.

File: pkg/cmd/validation/uuid.go (hunk 6 above)
Comment (mcncl) [reply pkg/cmd/validation/uuid.go:6]: Hmm, my personal idea with the CLI is that it reduces/entirely removes the requirement to go to the web. I don't really see much value in a CLI in which you need to go to the web to get the values.

**Comment by mcncl** [comment]
@jradtilbrook the table output feels like a preference over a function? Our API outputs JSON when a REST call is made, other tools such as `aws cli` output JSON when you `describe` a _thing_. A table is "nice" if the only output is the agent name and ID, but should data be masked if they opt to not select JSON or YAML?

**Comment by sj26** [comment]
Nice!

Instead of `view` may I suggest `get` to mirror the command structure from the agent:

https://github.com/buildkite/agent/blob/main/clicommand/meta_data_get.go

**Comment by mcncl** [comment]
@sj26 that feels like a fair suggestion and I'm all for consistency in the experience across our tooling. I've made that change.

Thank you!

**Comment by sj26** [comment]
(Love to see the CLI getting some love!)

**Comment by jradtilbrook** [comment]
This has been implemented although it doesnt have the output formatting which we will come back to later

**Pull request closed without merging**